COOP/List.h

```diff
@@ -0,0 +1,128 @@
+#ifndef __COOP__LIST__H_
+#define __COOP__LIST__H_
+
+#include "COOP.h"
+#include "Iterator.h"
+#include "SharedObjPtr.h"
+#include <stddef.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+/* ===== Element type enum for printing/formatting ===== */
+typedef enum {
+    LIST_ELEM_INT,
+    LIST_ELEM_FLOAT,
+    LIST_ELEM_CHAR,
+    LIST_ELEM_OBJ_SPTR,
+    LIST_ELEM_RAW_BYTES   /* default: hex dump of elementSize bytes */
+} List_ElementType;
+
+
+/* ===== Node ===== */
+DEF_CLASS(ListNode);
+ListNode* prev;
+ListNode* next;
+unsigned char payload[];   /* element bytes (length = elementSize) */
+END_DEF(ListNode);
+
+/* ===== Iterator derived from Iterator =====*/
+DEF_DERIVED_CLASS(ListIter, Iterator);
+void* list;
+ListNode* node;
+END_DEF_DERIVED(ListIter);
+
+/* ===== GenericList (homogeneous by elementSize) ===== */
+DEF_CLASS(GenericList);
+MEM_SIZE_T size;           /* number of elements */
+MEM_SIZE_T elementSize;    /* size of each element in bytes */
+ListNode* head;            /* first node */
+ListNode* tail;            /* last node */
+List_ElementType elem_type;
+ListIter begin_iter;
+ListIter end_iter;
+END_DEF(GenericList);
+
+
+
+#define LIST_UPDATE_ITERS_TAILEND(self)       \
+        (self)->begin_iter.list = (self);   \
+        (self)->begin_iter.node = (self)->head ? (self)->head : NULL; \
```

Comment:
```suggestion
        (self)->begin_iter.node = (self)->head; \
```
why it isn't enough?